Initial IRC Nick Modes Support

llmII opened the pull request:
Allows for each IRC Connection made by the relay to have a set of modes set or removed on the connection (can help with PM floods/spam).

qaisjp commented on bridge/bridge.go:
```diff
@@ -22,7 +22,8 @@ type Config struct {
 	IRCServerPass    string
-	IRCListenerName  string // i.e, "DiscordBot", required to listen for messages in all cases
+	IRCListenerName  string            // i.e, "DiscordBot", required to listen for messages in all cases
+	IRCNickModes     map[string]string // Modes to set on each IRC connection @OnWelcome
```
Have you tried an anonymous struct here, like this?

```
IRCNickModes struct {
    Add string
    Remove string
}
```

llmII replied:
I'll look into this soonish. I've got several other feature branches I need to clean up and submit pull requests for but after that's done I'll check if viper will do it and if so make use thereof.

llmII replied:
[Unmarshalling](https://github.com/spf13/viper#unmarshaling) is what you mean to look at? I would need to read more deeply and probably try some things but I get the feeling it would be applied to the entirety of the configuration... not just a single part?  If it has to be done for the entire config I believe that such an effort should be made into a separate pull request.

llmII replied:
Are we sure we want an anonymous struct, started work on this but you have to do the anonymous struct definition both in main.go and bridge.go files and seems like it's code duplication at that point. Can understand wanting unmarshaling and could still do that but not sure if an anonymous struct is the right way to go for it here.

llmII replied:
Seems an anonymous struct won't work:
```
./main.go:76:42: cannot use &ircNickModes (type *struct { Add string "mapstructure:\"add\""; Del string "mapstructure:\"del\"" }) as type viper.DecoderConfigOption in argument to viper.Unmarshal`
```
Going to use a regular struct.

llmII replied:
I was wrong, called the wrong function. it now uses an anonymous struct.

qaisjp commented:
Hmmm, do you think this feature could be replaced with a `irc_puppet_prejoin_commands` setting, with support for `${NICK}` interpolation? (i.e. similar to #71 but a config option that sends for puppets only)

llmII commented:
> Hmmm, do you think this feature could be replaced with a `irc_puppet_prejoin_commands` setting, with support for `${NICK}` interpolation? (i.e. similar to #71 but a config option that sends for puppets only)

That's actually why I split it out because I couldn't ever come up with a reason for the listener to apply modes to itself and to keep from having a specialization within the `irc_puppet_prejoin_commands` setting. That said interpolation within `irc_puppet_prejoin_commands` probably isn't a bad idea. If we added interpolation support I'd probably want to keep the modes setting separated as well for connections that aren't listener and allow for the perform to actually set modes on the listener if someone were to have in the list "MODE ${NICK} +abcd-efg" or whatnot. Just means that that would be for modes shared among all clients spawned vs the modes that are specifically to be set for non-listener connections.

qaisjp commented:
To clarify, the config option for _all_ connections is `irc_prejoin_commands`, and I'm suggesting we add a `irc_puppet_prejoin_commands` option just for the puppet (non-listener) connection.

I would definitely make use of this feature. At the moment I hardcoded it to `+D` because that's what I needed, but we could just move that to a default setting like `irc_puppet_prejoin_commands: ["MODE ${NICK} +D"]`

llmII commented:
> To clarify, the config option for _all_ connections is `irc_prejoin_commands`, and I'm suggesting we add a `irc_puppet_prejoin_commands` option just for the puppet (non-listener) connection.
> 
> I would definitely make use of this feature. At the moment I hardcoded it to `+D` because that's what I needed, but we could just move that to a default setting like `irc_puppet_prejoin_commands: ["MODE ${NICK} +D"]`

That's a far better solution actually, an `irc_prejoin_commands` for listener and `irc_puppet_prejoin_commands` for puppets. Now that you mention it I'm already using the modes thing from before to handle unsetting +D and setting +d (inspircd makes +D privdef and +d regular deaf). I'll put this on my list to split it up and strip out any hardcoded mode sets.

llmII commented:
PR #88 supplants this, going to close (keeping branch around in case for some reason the new PR proves insufficient).
